packages/itwin/tree-widget/src/components/trees/stateless/models-tree/HierarchyVisibilityHandler.ts

```diff
@@ -0,0 +1,877 @@
+        defer(() => {
+          this.changeCategoryStateInViewportAccordingToModelVisibility(modelId, categoryId, on);
+          return this.clearAlwaysAndNeverDrawnElements(props);
+        }),
+      );
+    });
+
+    const ovr = this._props.overrides?.changeCategoryState;
+    return ovr ? from(ovr(this.createVoidOverrideProps(props, result))) : result;
+  }
+
+  /**
+   * Updates visibility of all grouping node's elements.
+   * @see `changeElementState`
+   */
+  private changeElementGroupingNodeState(node: GroupingHierarchyNode, on: boolean): Observable<void> {
+    const result = this.getGroupedElementIds(node).pipe(
+      mergeMap((group) => {
+        if (!group) {
+          return EMPTY;
+        }
+        const { modelId, categoryId, elementIds } = group;
+        const viewport = this._props.viewport;
+        return concat(
+          on && !viewport.view.viewsModel(modelId) ? from(viewport.addViewedModels(modelId)) : EMPTY,
+          defer(() => {
+            const categoryVisibility = this.getDefaultCategoryVisibilityStatus(categoryId, modelId);
+            const isDisplayedByDefault = categoryVisibility.state === "visible";
+            return from(elementIds).pipe(this.changeElementStateNoChildrenOperator({ on, isDisplayedByDefault }));
+          }),
+        );
+      }),
+    );
+
+    const ovr = this._props.overrides?.changeElementGroupingNodeState;
+    return ovr ? from(ovr(this.createVoidOverrideProps({ node, on }, result))) : result;
+  }
+
+  /**
+   * Updates visibility of an element and all its child elements by adding them to the always/never drawn list.
+   * @note If element is to be enabled and model is hidden, it will be enabled.
+   */
+  private changeElementState(props: ChangeElementStateProps): Observable<void> {
+    const result = defer(() => {
+      const { elementId, on, hasChildren, modelId, categoryId } = props;
+      const viewport = this._props.viewport;
+      return concat(
+        props.on && !viewport.view.viewsModel(modelId) ? this.showModelWithoutAnyCategoriesOrElements(modelId) : EMPTY,
+        defer(() => {
+          const categoryVisibility = this.getDefaultCategoryVisibilityStatus(categoryId, modelId);
+          const isDisplayedByDefault = categoryVisibility.state === "visible";
+          return of(elementId).pipe(
+            concatWith(hasChildren ? from(this._queryHandler.queryElements({ rootElementIds: props.elementId })) : EMPTY),
+            this.changeElementStateNoChildrenOperator({ on, isDisplayedByDefault }),
+          );
+        }),
+      );
+    });
+
+    const ovr = this._props.overrides?.changeElementState;
+    return ovr ? from(ovr(this.createVoidOverrideProps(props, result))) : result;
+  }
+
+  private changeElementStateNoChildrenOperator(props: { on: boolean; isDisplayedByDefault: boolean }): OperatorFunction<string, void> {
+    return (elementIds: Observable<string>) => {
```

Comment:
ids should use `Id64String` type rather than `string`

Reply:
Fixed. P.S I wish regular string wasn't assignable to Id64String. It could be implemented as a pattern type for example.

Reply:
I guess they could, to provide a little more type safety, e.g. `type Id64String = `0x${string}`. However, it seems that would be a fairly large breaking change in `itwinjs-core`.